release.py

```diff
@@ -0,0 +1,289 @@
+#! /usr/bin/env python
+# Tahoe-LAFS -- secure, distributed storage grid
+#
+# Copyright © 2006-2012 The Tahoe-LAFS Software Foundation
+#
+# This file is part of Tahoe-LAFS.
+#
+# See the docs/about.rst file for licensing information.
+import os
+import shutil
+import re
+import sys
+import subprocess
+import argparse
+import datetime
+import pathlib
+
+
+class bcolors:
+    HEADER = "\033[95m"
+    OKBLUE = "\033[94m"
+    OKCYAN = "\033[96m"
+    OKGREEN = "\033[92m"
+    WARNING = "\033[93m"
+    FAIL = "\033[91m"
+    ENDC = "\033[0m"
+    BOLD = "\033[1m"
+    UNDERLINE = "\033[4m"
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--clean", action="store_true", help="Cancel existing release process, clean files"
+)
+parser.add_argument(
+    "--ignore-deps", action="store_true", help="Ignore dependency checks"
+)
+parser.add_argument("--fin", action="store_true", help="Finish release proccess")
+parser.add_argument("--sign", type=str, help="Signing key")
+parser.add_argument("--ticket", type=int, help="Ticket number", required=True)
+parser.add_argument(
+    "--repo",
+    type=str,
+    help="Destination repo (fork), example: git@github.com:meejah/tahoe-lafs.git",
+)
+parser.add_argument("--retry", action="store_true", help="Retry release")
+parser.add_argument("--tag", type=str, help="Release tag", required=True)
+args = parser.parse_args()
+
+
+TAG = args.tag
+TICKET = args.ticket
+TODAY = datetime.date.today()
+BRANCH = "{ticket}.release-{tag}".format(
+    ticket=TICKET, tag=TAG
+)  # looks like XXXX.release-1.16.0
+RELEASE_TITLE = "Release {tag} ({date})".format(tag=TAG, date=TODAY)
+RELEASE_FOLDER = pathlib.Path("../tahoe-release-{0}".format(TAG))
+RELEASE_PROGRESS = pathlib.Path("../.tahoe-release-{0}-progress".format(TAG))
+CONTINUE_INSTRUCTION = (
+    bcolors.BOLD
+    + "Instruction : run ./venv/bin/python release.py --ignore-deps --tag {tag} --ticket {ticket} --sign {sign} --fin".format(
+        tag=TAG, ticket=TICKET, sign="YOUR_SIGNING_KEY_HERE"
+    )
+    + bcolors.ENDC
+)
+
+
+def clean():
+    shutil.rmtree(RELEASE_FOLDER, ignore_errors=True)
+    shutil.rmtree(RELEASE_PROGRESS, ignore_errors=True)
+
+
+def check_dependencies():
+    repo_clean = subprocess.run(["git", "status", "-s"], capture_output=True)
```

Comment:
This check isn't correct; if there are any unknown files it fails.
It should only fail if we have changes to existing files.

(And, actually, why does it even matter? Aren't the operations performed in a fresh checkout?)